### ISSUE TITLE
Counter Reset Aggregates

### ISSUE THREAD

**Issue by davidkohn88** [timescale/timescaledb-toolkit#4]
Keeping track of counters is a common timeseries task, these are continuously increasing values that may at some times reset themselves, ie, when they reach say INTMAX or when a device reboots etc. Keeping track of these resets is a bit of a difficult problem as it requires arranging the values in time order and then noting when a decrease has occurred, whenever a decrease occurs it is assumed that the counter was reset and the previous value should be added to future values (or some static value should be added to future values as would be the case if the counter always reset at INTMAX). 

For example: 
```
         tstamp         | raw_counter | adjusted_counter
------------------------+-------------+------------------
 2020-12-02 21:47:50+00 |          10 |               10
 2020-12-02 21:57:50+00 |         120 |              120
 2020-12-02 22:07:50+00 |        3045 |             3045
 2020-12-02 22:17:50+00 |         205 |             3250 // <- Counter reset
 2020-12-02 22:27:50+00 |        5200 |             8450
 2020-12-02 22:37:50+00 |         500 |             8950 // <- Counter reset
 2020-12-02 22:47:50+00 |        3010 |            11960
```
Shows a typical counter reset in which final value of the counter before the reset is then added to successive values in order get the true value of the counter at successive times. The true value of the counter is not necessarily what is important at each point (for instance, perhaps there was a reset of the counter sometime before this that we don't know about because it's outside of the scope of our analysis), but rather the difference between the counter's value at the beginning and the end of the period, in this case, `11960-10 = 11950` is the important characteristic, however, given the possibility of resets, the values at each point in between must be known in order to calculate it accurately. 

This type of calculation leads to some problems in terms of efficient storage and calculation with counter values in Postgres/TimescaleDB: 
- A typical aggregate will not work to calculate all sorts of counter values, you either need to use an ordered set aggregate or, more typically a window function in order to perform analyses on counters. 
- Window functions and ordered set aggregates are not parallelizable in Postgres, which means one of the main mechanisms for speeding up analyses will not work, and any parallelization that does not take into account time will give wrong results.
- The typical approach also means we cannot incorporate counters into continuous aggregates in TimescaleDB because they are not partializable and combinable. Which means one of the main ways to speed up intensive calculations is also out. 
- The typical approach cannot easily be pushed down to data nodes in a multi-node setup so that more of the work is kept on the data node rather than access node (ie partitionwise aggregation)

So, the question is, can we devise an aggregate that gets around some of these limitations in order to make calculations on counters more performant and able to be integrated into continuous aggregates? 

Related issues: https://github.com/timescale/timescaledb/issues/2479 (there were multiple other customer issues we ran into where we provided workarounds outside of github, this is all I could find in a brief search on github though)

**Comment by JLockerman** [comment]
A preliminary version of the counter aggregate was released in 0.2.0 and should be available on Forge at the early next week.

**Comment by epgts** [comment]
I accidentally reopened this on February 22. Closing again.
